magpie/services.py

```diff
@@ -298,20 +304,25 @@ def effective_permissions(self, user, resource, permissions=None, allow_match=Tr
                     if not match and perm_set.scope == Scope.MATCH:
                         continue
 
+                    # less obvious use case for both of the following user/group blocks:
+                    #   no need to check explicitly for ALLOW since it was either already set during previous iteration
+                    #   (at that moment, perm=None) or a DENY was already set, but it takes precedence over it anyway
+
                     # user direct permissions have priority over inherited ones from groups
                     if perm_set.type == PermissionType.DIRECT:
                         perm = effective_perms.get(perm_name)
-                        # explicit user deny overrides user allow if any already found
+                        # explicit user DENY overrides user ALLOW if any already found
+                        # if permission name not already found, ALLOW/DENY is set regardless (first occurrence)
                         if perm is None or perm.type == PermissionType.INHERITED or perm_set.access == Access.DENY:
                             effective_perms[perm_name] = perm_set
                         continue  # final decision for this user, skip any group permissions
 
-                    # otherwise check if for another group permission and yet no user permission
+                    # otherwise check for group permission
+                    # like previously, explicit DENY overrides ALLOW if permission name was already found
+                    # if permission name not already found, ALLOW/DENY is set regardless (first occurrence)
                     perm = effective_perms.get(perm_name)
-                    if perm is None or perm.type == PermissionType.INHERITED:
-                        # explicit deny overrides allow if any already found
-                        if perm is None or perm_set.access == Access.DENY:
-                            effective_perms[perm_name] = perm_set
+                    if perm is None or (perm.type == PermissionType.INHERITED and perm_set.access == Access.DENY):
```

Comment:
Thanks for the update, much easier to follow. I think it still miss a point about the perm.type == PermissionType.INHERITED.
If the perm is direct (line 312), an allow permission overrides any access type that has been inherited. It means that a direct "allow" takes precedance over inherited "deny",
Here, to override any inherited permissions the access type must be deny. It means that a "deny" group permission take precedance over other "allow" group permission.
In short, we should better explain the difference between INHERITED  or DENY of line 316 vs INHERITED and DENY of line 324

Reply:
I feel that is what comments on lines 311 and 318 describe,

Reply:
I'll try to be more explicit. The following 3 rules should be clear while reading the comments:
1. direct > inherited (line 311 is fine and line 318 cover the fact that once direct is set we're done, but line 314 put emphasis on DENY override ALLOW while opposite is true too (the reason why we got a type == inherited **or** access == DENY is not covered) 
2. direct DENY > direct ALLOW  (line 314)
3. inherited DENY > inherited ALLOW (line 321 is confusing because it's stating **like previously** while this time it's between group (the reason why we got a type == inherited **and** access == DENY is not covered)

Reply:
Maybe it is because of the way I parse the sequence that doesn't feel ambiguous for me. 

1. User > Group - so if Group was set (ALLOW or DENY) we don't care, we set whatever User permission says. 
   For this reason, the opposite use case is never true.
2. We then enter User branch. We are in the realm of User-only permissions. So DENY > ALLOW is the only thing to evaluate.
----
3. ONLY THEN, once we established we are not evaluating User permissions, we compare only Group with Group. 
4. Again, the only thing to check is DENY > ALLOW resolution. 

It is mostly for this reason I had 2 ifs before. It gave a clean distinction between the 2 phases (eval User > Group) and then, for each (DENY > ALLOW). 
If you think about it, the *original* 2nd if of the Group block is identical to the 2nd one of the User group 
`if perm is None or perm.type == PermissionType.INHERITED or perm_set.access == Access.DENY:`
Only thing is that we have a redundant check for `perm.type == PermissionType.INHERITED` since we evaluated that on the 1st if of the Group block, which is why it became `if perm is None or perm_set.access == Access.DENY:`

Making the if all-in-one for Group portion makes it more complicated to evaluate in my opinion because it introduces the mixing of Type and Access simultaneously.

Reply:
> 1. [...] For this reason, the opposite use case is never true.

I meant if group was DENY, but user is ALLOW, allow override deny. Just explaining that would explain perfectly why we have a inherited or deny. For me line 311, 314 and 318 doesn't make that obvious.

> 2. [...] So DENY > ALLOW is the only thing to evaluate. 

This statement cover the second part of the if of l316. And the comment just above cover only that part. My point is reallly to cover the inherited part too. Comment should look like that : "permission is set if not already found (first occurence), if inherited regardless of access (direct > inherited) or deny (deny > allow)"

> 3 [...] we compare only Group with Group

Exact. But comment said "like previously" which compare direct with group and direct with direct.
I would change the comment to put the emphasis on the group with group **and** deny > allow to reflect the if, something like : "otherwise check for group permission, permission is set if not already found (first occurence) or if inherited (compare group with group only) AND deny (deny > allow)"

Sorry to be bothering, but I'm only looking to have better comments explaining what is going on for non-initiated, not to refactor the whole part.